packages/utils/src/logger.ts

```diff
@@ -0,0 +1,396 @@
+ * - Console styling support
+ *
+ * @example Typical usage - disabled by default with debug level
+ * ```typescript
+ * import { getLogger } from "@commontools/utils/logger";
+ *
+ * // Common pattern: create a debug logger that's disabled by default
+ * // pass in function for lazy evaluation of parameters
+ * const logger = getLogger({ enabled: false, level: "debug" });
+ * logger.debug(() => ["Processing:", data]);
+ * ```
+ *
+ * @example Basic usage
+ * ```typescript
+ * import { log, setLogLevel } from "@commontools/utils/logger";
+ *
+ * // Simple logging (uses info level by default)
+ * log("Application started");
+ *
+ * // Change global log level
+ * setLogLevel("debug");
+ * ```
+ *
+ * @example Module-tagged logging
+ * ```typescript
+ * import { getLogger } from "@commontools/utils/logger";
+ *
+ * // Auto-detects calling module from stack trace - recommended approach
+ * const logger = getLogger();
+ *
+ * // Logs will show: [INFO][your-module::HH:MM:SS.mmm] message
+ * logger.info("Processing user data");
+ * logger.debug("Cache hit for user", userId);
+ * logger.warn("API rate limit approaching");
+ * logger.error("Failed to save user", error);
+ * ```
+ *
+ * @example Lazy evaluation for expensive operations
+ * ```typescript
+ * const logger = getLogger();
+ *
+ * // Function is only called if debug level is active
+ * logger.debug(() => `Computed value: ${expensiveComputation()}`);
+ *
+ * // Works with arrays that get flattened
+ * logger.info(() => ["Processing", count, "items"]);
+ * ```
+ *
+ * @example Per-logger configuration
+ * ```typescript
+ * // Create a debug logger for development
+ * const debugLogger = getLogger({
+ *   level: "debug",  // Show all messages for this logger
+ *   enabled: true    // Explicitly enable
+ * });
+ *
+ * // Create a disabled logger for verbose sections
+ * const verboseLogger = getLogger({ enabled: false });
+ *
+ * // Enable/disable at runtime
+ * verboseLogger.disabled = false; // Now it will log
+ * verboseLogger.info("This will show");
+ * ```
+ */
+
+export type LogMessage = unknown | (() => unknown);
+
+export type LogLevel = "debug" | "info" | "warn" | "error";
+
+/**
+ * Numeric values for log levels to enable comparison
+ */
+const LOG_LEVELS: Record<LogLevel, number> = {
+  debug: 0,
+  info: 1,
+  warn: 2,
+  error: 3,
+};
+
+/**
+ * Colors for each log level
+ */
+export const LOG_COLORS = {
+  debug: "color: #6b7280",
+  info: "color: #6b7280",
+  warn: "color: #eab308",
+  error: "color: #ef4444",
+  // Tagged logger colors
+  taggedDebug: "color: #6b7280; font-weight: 500",
+  taggedInfo: "color: #10b981; font-weight: 500",
+  taggedWarn: "color: #eab308; font-weight: 500",
+  taggedError: "color: #ef4444; font-weight: 500",
+} as const;
+
+/**
+ * Current minimum log level - messages below this level are ignored
+ */
+let currentLogLevel: LogLevel = "info";
+
+/**
+ * Initialize log level from environment variable if available
+ */
+if (typeof Deno !== "undefined") {
```

Comment:
`deno-web-test` is an environment that has `globalThis.Deno` but not `Deno.env` -- I think `@commontools/utils/env` handles this distinction